### ISSUE TITLE
Support a PPO/TRPO example for continuous control problem

### ISSUE THREAD

**Issue by dmuestc** [reinforceio/tensorforce#284]
Hi，I want to run some experiments for reproducing paper's results with tensorforce, mainly related to continuous control problem using PPO/TRPO. I tried different configurations many times, but the results didn't converge, I doubt something wrong with my configurations or something else, this is my config:

```json
env=Pendulum

{
    "type": "ppo_agent",
    "max_episodes": 5000,
    "max_episode_timesteps": 200,

    "batch_size": 1024,
    "batched_observe": 1024,
    "step_optimizer": {
        "type": "adam",
        "learning_rate": 5e-3
    },
    "optimization_steps": 10,

    "discount": 0.97,

    "entropy_regularization": 1e-2,
    "gae_lambda": 0.01,
    "likelihood_ratio_clipping": 0.2,

    "baseline_mode": "states",
    "baseline": {
      "type": "mlp",
      "sizes": [64, 64]
    },
    "baseline_optimizer": {
        "type": "multi_step",
        "optimizer": {
            "type": "adam",
            "learning_rate": 5e-4
        },
        "num_steps": 5
    },

    "network": [
        {
            "type": "dense",
            "size": 64,
            "activation": "tanh"
        },
        {
            "type": "dense",
            "size": 64,
            "activation": "tanh"
        }
    ]
}
```
The given examples for PPO/TRPO mainly are CartPole-v0, could you help me to give a few continuous examples？

**Comment by AlexKuhnle** [comment]
Hey @dmuestc,

the config looks alright, two things to be aware though: First, Pendulum uses bounded continuous actions, where TensorForce implicitly uses the Beta distribution (unless explicitly configured otherwise), which probably is different to some papers. Worth trying bigger learning rates, which in our experience seems to help training when using the Beta distribution. Second, the `multi_step` optimizer is only a currently-best-possible approximation to the optimization procedure described in the PPO paper (multiple optimization steps for a batch, instead of subsampling smaller batches). Similar to moving the optimizers with the last update, we are working on moving the batching/memory to TensorFlow, which will allow to resemble PPO more closely again.

**Comment by dmuestc** [comment]
Thanks @AlexKuhnle reply, I'll follow you suggestions and try again.

**Comment by dmuestc** [comment]
hi, @AlexKuhnle, I have another question about the detail about GAE, here:
https://github.com/reinforceio/tensorforce/blob/fe858269d020e3a18ec92332ee7417aded9b7ec1/tensorforce/models/pg_model.py#L168

I suspect there is a problem with such implementation,  my idea is that the sequence terminals should not be applied in "state_value_change" but in the "next state_value", here is my implementation:
```python
 terminal = tf.concat(values=(terminal, (False,)), axis=0)
 state_value = tf.concat(values=(state_value, (0.0,)), axis=0)
 td_residual = reward - state_value[:-1] + state_value[1:]*self.discount*(1.0-tf.cast(terminal[1:], tf.float32))
 gae_discount = self.discount * self.gae_lambda
 reward = self.fn_discounted_cumulative_reward(terminal=terminal[1:], reward=td_residual, discount=gae_discount)
```
And I run some experiments in tensorforce-benchmark, like vpg, ppo, and trpo, it seems more stable than before.

**Comment by AlexKuhnle** [comment]
So what you say -- and as far as I can tell the only difference -- is that if `s_t` is terminal, then `A(s_t) = r_t` in our version so far, but you suggest it should be `A(s_t) = r_t - V(s_t)`. That sounds correct to me, well spotted! We will fix this soon, thanks for that. I'm a bit surprised that this causes a recognizable difference in performance, but the better, performance improvements are always very welcome... :-)

**Comment by AlexKuhnle** [comment]
The following should fix our implementation in the way you suggested:
```
zeros = tf.zeros_like(tensor=state_value)
next_state_value = tf.where(condition=terminal, x=zeros, y=state_value)
state_value_change = self.discount * next_state_value[1:] - state_value[:-1]
state_value_change = tf.concat(values=(state_value_change, (0.0,)), axis=0)
td_residual = reward + state_value_change
```

**Comment by AlexKuhnle** [comment]
Sorry, this is better:
```
zeros = tf.zeros_like(tensor=state_value)
next_state_value = tf.where(condition=terminal, x=zeros, y=state_value)
next_state_value = tf.concat(values=(next_state_value, (0.0,)), axis=0)
state_value_change = self.discount * next_state_value[1:] - state_value
td_residual = reward + state_value_change
```

**Comment by michaelschaarschmidt** [comment]
Implemented change with https://github.com/reinforceio/tensorforce/commit/2de12dc289bf5635ddc5855a3deedbbe2621a2ff

**Comment by michaelschaarschmidt** [comment]
@dmuestc if you have identified any well working configurations, a PR to https://github.com/reinforceio/tensorforce-benchmark is very welcome, where we will try to collect more

**Comment by dmuestc** [comment]
Thanks @michaelschaarschmidt.

As far as I know, there is still a bit difference with the true GAE, because here we assume the next state is terminal state of the last sample, namely` _V_(s_t+1)` is zero, maybe it's more reasonable to drop the last sample in the batch in current version. Or do a fix in the later--especially for probably incomplete episode in a batch-- as well as the `tf_discounted_reward`.

**Comment by dmuestc** [comment]
@michaelschaarschmidt, sorry, but maybe the terminal should be [terminal[1:], True], here:
https://github.com/reinforceio/tensorforce/blob/2de12dc289bf5635ddc5855a3deedbbe2621a2ff/tensorforce/models/pg_model.py#L174

**Comment by AlexKuhnle** [comment]
You're right in that timestep-based updates are generally not ideal for policy gradient models, which estimate cumulative reward using the experienced reward of an entire episode. We're working on memories and batching right now, and one goal is to offer an episode-based batching mode. Instead of a batch size of, say, 1000 (implicitly timesteps) you can then specify a batch size of 10 episodes, for instance. Another option will probably be to have batches with instances of the form S,A,R,S_next (or sequences S,A,R,S,...,R,S_next) which would at least allow to estimate the reward of the episode via the baseline (if available), similar to Q models.

**Comment by AlexKuhnle** [comment]
Why do you think the terminal shift is necessary?

**Comment by dmuestc** [comment]
here is the idea, if use for loop, is:
```python
last = 0
for t in reversed(range(SEQ_LEN)):
    td = R[t] + gamma * V[t+1] * (1-terminal[t+1]) - V[t]
    GAE[t] = last = td + gamma * lambda * (1-terminal[t+1]) * last
```
and in the last line, we actually calculate the discounted td, and init GAE give zero, so the terminal should shift a step, like: 
    `G[t] = TD[t] + G[t+1] * discount * terminal[t+1] `
and here for terminal, step `t+1` should be used not step `t`.
because the last terminal is undefined, so I append a bool value "True",

**Comment by AlexKuhnle** [comment]
I see, so that would be better:
```
next_state_value = tf.concat(values=(state_value[1:], (0.0,)), axis=0)
zeros = tf.zeros_like(tensor=next_state_value)
next_state_value = tf.where(condition=terminal, x=zeros, y=next_state_value)
state_value_change = self.discount * next_state_value[1:] - state_value
td_residual = reward + state_value_change
```

**Comment by dmuestc** [comment]
Thanks a lot. I'm trying to do a some change in current version, such as sample a mini-batch from that big batch, but find it's a bit hard. Because I think I need take a lot time to read the code, the structure and abstraction of the tensorforce is such sophisticated!  :-)
When do we will see the new all-graph version, I'm sincerely looking forward it!

**Comment by AlexKuhnle** [comment]
As always, it takes longer than expected, and I expect, let's say, two weeks... so in about a month? :-)  Seriously, it's hard to say, particularly with limited time, but it's in the making, and ideally we push it as a separate branch soon...

**Comment by lefnire** [comment]
FWIW I saw a perf boost on avg across my PPO runs since 2de12dc289bf5635ddc5855a3deedbbe2621a2ff
